spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java

```diff
@@ -51,52 +71,77 @@ public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFact
 	public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFactory factory,
 			@Nullable Sinks.Empty<Void> completionSink) {
 
-		super(session, ObjectUtils.getIdentityHexString(session), info, factory, completionSink);
-		// TODO: suspend causes failures if invoked at this stage
-		// suspendReceiving();
+		super(session, ObjectUtils.getIdentityHexString(session), info, factory);
+		this.handlerCompletionSink = completionSink;
+		this.flux = Flux.create(emitter -> {
+			this.sink = emitter;
+			emitter.onRequest(n -> {
+				boolean demand = false;
+				this.lock.lock();
+				try {
+					this.requested += n;
+					if (!this.awaitingMessage && this.requested > 0) {
+						this.requested--;
+						this.awaitingMessage = true;
+						demand = true;
+					}
+				}
+				finally {
+					this.lock.unlock();
+				}
+
+				if (demand) {
+					getDelegate().demand();
+				}
+			});
+		});
 	}
 
+	void handleMessage(WebSocketMessage message) {
+		this.sink.next(message);
 
-	@Override
-	protected boolean canSuspendReceiving() {
-		// Jetty 12 TODO: research suspend functionality in Jetty 12
-		return false;
+		boolean demand = false;
+		this.lock.lock();
+		try {
+			if (!this.awaitingMessage) {
+				throw new IllegalStateException();
+			}
+			this.awaitingMessage = false;
+			if (this.requested > 0) {
+				this.requested--;
```

Comment:
same as above, this should take the "unbounded request" scenario (in the Reactive Streams sense) into consideration

Reply:
done.